lib/active_storage/service/s3_service.rb

```diff
@@ -4,8 +4,14 @@
 class ActiveStorage::Service::S3Service < ActiveStorage::Service
   attr_reader :client, :bucket
 
-  def initialize(access_key_id:, secret_access_key:, region:, bucket:)
-    @client = Aws::S3::Resource.new(access_key_id: access_key_id, secret_access_key: secret_access_key, region: region)
+  def initialize(access_key_id:, secret_access_key:, region:, bucket:, endpoint: nil)
+    params = {
+      access_key_id:     access_key_id,
+      secret_access_key: secret_access_key,
+      region:            region
+    }
+    params.merge!(endpoint ? {endpoint: endpoint, force_path_style: true} : {})
```

Comment:
Last comment, I promise: leave a space after the opening curly brace and before the closing one:

```ruby
{ endpoint: endpoint, force_path_style: true }
```